### PR TITLE
Update debian

### LINKED ISSUE #123: update dev container
You can assign this issue to me
update development container to use debian as close as possible to the one of the production containers

### PR DESCRIPTION
**Pull request by ddavidebor**
Fixes #123 
Partially addresses #134 

So as you can see from the changenote I've:

- Moved the image from debian to the same minideb used by the bitnami images
- Used the install_packages script from bitnami
- Moved python and nvm/node to the frappe user

I'd recommend testing manually

### CODE REVIEW

File: build/bench/Dockerfile [hunk 1]
```diff
@@ -1,59 +1,114 @@
...
 
 # Install wkhtmltox correctly
 RUN wget https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.5/wkhtmltox_0.12.5-1.stretch_amd64.deb
 RUN dpkg -i wkhtmltox_0.12.5-1.stretch_amd64.deb && rm wkhtmltox_0.12.5-1.stretch_amd64.deb
 
-# Install bench
-RUN pip3 install -e git+https://github.com/frappe/bench.git#egg=bench --no-cache
-
+# Create new user with directory, add user to sudo group, allow passwordless sudo, switch to that user and change directory to user home directory
+RUN useradd --no-log-init -r -m -g  sudo frappe
```
Comment (revant) [inline build/bench/Dockerfile]: why does frappe need sudo?

I removed it and managed to get everything working.

sudo is required mainly to restart supervisor, upgrade letsencrypt. All of that is not required in this docker image. I don't think sudo is required for development

File: build/bench/Dockerfile [hunk 2]
```diff
@@ -1,59 +1,114 @@
...
+ENV NVM_DIR=/home/frappe/.nvm
+RUN . "$NVM_DIR/nvm.sh" && nvm install ${NODE_VERSION}
+RUN . "$NVM_DIR/nvm.sh" && nvm install ${NODE_VERSION_FRAPPEV11}
+RUN . "$NVM_DIR/nvm.sh" && nvm use v${NODE_VERSION}
+RUN . "$NVM_DIR/nvm.sh" && nvm alias default v${NODE_VERSION}
+# Add NVM path for user in Dockerfile, interactive shell is already managed by the NVM install script
+ENV PATH="/home/frappe/.nvm/versions/node/v${NODE_VERSION}/bin/:${PATH}"
+
+# Install yarn
+RUN npm install -g yarn
```
Comment (revant) [inline build/bench/Dockerfile]: If possible reduce the layers with common `RUN`

File: build/bench/Dockerfile (hunk 1 above)
Comment (ddavidebor) [reply build/bench/Dockerfile]: In my case I found it necessary if the user wants to install additional packages using apt-get

File: build/bench/Dockerfile (hunk 2 above)
Comment (ddavidebor) [reply build/bench/Dockerfile]: Done

File: build/bench/Dockerfile (hunk 1 above)
Comment (revant) [reply build/bench/Dockerfile]: I'll prefer not to have sudo.

If root access is needed. It is container anyway. just `docker exec -u 0 ...`

having sudo will encourage blind use of sudo. "If it doesn't work, use sudo"

**Comment by revant** [comment]
I'll prefer not to have sudo.

If root access is needed. It is container anyway. just `docker exec -u 0 ...`

having sudo will encourage blind use of sudo. "If it doesn't work, use sudo"

**Comment by ddavidebor** [comment]
> I'll prefer not to have sudo.
> 
> If root access is needed. It is container anyway. just `docker exec -u 0 ...`

That's rather impractical once you are inside the VScode environment, working inside the container. 

I am 100% with you on a production image, but for this, I'm really worried about having an easy getting-started and a workflow that mimics what people are used to. This has really been a friction point with the frappe framework in the past, and I'd like to address that. 

But I could easily write a sudo warning message that tells people exactly for what they should and should not use sudo in this container. That would be a bit softer on the less experienced programmers.

File: build/bench/Dockerfile [hunk 3]
```diff
@@ -1,59 +1,112 @@
...
+RUN bash -c "node --version" \
+    && bash -c "npm --version" \
+    && bash -c "yarn --version"
+
+
+WORKDIR /home/frappe/frappe-bench
 EXPOSE 8000 9000 6787
 
-VOLUME [ "/home/frappe/frappe-bench" ]
+VOLUME [ "/root/frappe-bench" ]
```
[the remark this answers is not in the document]
Comment (ddavidebor) [reply build/bench/Dockerfile]: fixed

**Comment by ddavidebor** [comment]
There seem to be something wrong with sudo regardless, passwordless sudo not working

**Comment by revant** [comment]
I built this image locally to test.

I'm getting permission denied while working in /workspace/development

permission-denied-bench

**Comment by ddavidebor** [comment]
Weird. What OS are you on?

**Comment by revant** [comment]
I tried to minimize layers. Layers can be reduced further. It will result in less layers to download.

Something like this :

<details>

```Dockerfile
# Frappe Bench Dockerfile
FROM bitnami/minideb:latest
LABEL author=frappé

RUN install_packages \
    git \
    wkhtmltopdf \
    mariadb-client \
    gettext-base \
    wget \
    # for PDF
    libssl-dev \
    fonts-cantarell \
    xfonts-75dpi \
    xfonts-base \
    # to work inside the container
    locales \
    build-essential \
    cron \
    curl \
    vim \
    sudo \
    iputils-ping \
    watch \
    tree \
    nano \
    software-properties-common \
    # For psycopg2
    libpq-dev \
    build-essential \
    # Other
    libffi-dev \
    liblcms2-dev \
    libldap2-dev \
    libmariadbclient-dev \
    libsasl2-dev \
    libtiff5-dev \
    libwebp-dev \
    redis-tools \
    rlwrap \
    tk8.6-dev \
    fonts-cantarell \
    # VSCode container requirements
    net-tools \
    # PYTHON
    python3-dev \
    python3-pip \
    python3-setuptools \
    python3-tk \
    python-virtualenv

# Install wkhtmltox correctly
RUN wget https://github.com/wkhtmltopdf/wkhtmltopdf/releases/download/0.12.5/wkhtmltox_0.12.5-1.stretch_amd64.deb
RUN dpkg -i wkhtmltox_0.12.5-1.stretch_amd64.deb && rm wkhtmltox_0.12.5-1.stretch_amd64.deb

# Create new user with directory, add user to sudo group, allow passwordless sudo, switch to that user and change directory to user home directory
RUN useradd --no-log-init -r -m -g  sudo frappe
RUN echo "frappe ALL=(ALL) NOPASSWD: ALL" >> /etc/sudoers
USER frappe
WORKDIR /home/frappe

# Install bench in the local user home directory
RUN pip3 install --user  git+https://github.com/frappe/bench.git#egg=bench --no-cache

# Export python executables for Dockerfile
ENV PATH=/home/frappe/.local/bin/:$PATH
# Export python executables for interactive shell
RUN echo "export PATH=/home/frappe/.local/bin/:\$PATH" >> /home/frappe/.bashrc \
# Print version and verify bashrc is properly sourced so that everything works in the Dockerfile
    && bench --version \
# Print version and verify bashrc is properly sourced so that everything works in the interactive shell
    && bash -c "bench --version"

# !!! UPDATE NODEJS PERIODICALLY WITH LATEST VERSIONS !!!
# https://nodejs.org/en/about/releases/
# https://nodejs.org/download/release/latest-v10.x/
# https://nodejs.org/download/release/latest-v12.x/
# https://nodejs.org/download/release/latest-v13.x/
ENV NODE_VERSION=12.16.1
ENV NODE_VERSION_FRAPPEV11=10.19.0

# Install nvm with node
RUN wget https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh
RUN chmod +x install.sh
RUN ./install.sh

ENV NVM_DIR=/home/frappe/.nvm
RUN . "$NVM_DIR/nvm.sh" \
    && nvm install ${NODE_VERSION_FRAPPEV11} \
    # Install yarn for nodejs v10
    && npm install -g yarn \
    && nvm install ${NODE_VERSION} \
    && nvm use v${NODE_VERSION} \
    # Install yarn for nodejs v12
    && npm install -g yarn \
    && nvm alias default v${NODE_VERSION}

# Add NVM path for user in Dockerfile, interactive shell is already managed by the NVM install script
ENV PATH="/home/frappe/.nvm/versions/node/v${NODE_VERSION}/bin/:${PATH}"

# Print version and verify bashrc is properly sourced so that everything works in the Dockerfile
RUN node --version \
    && npm --version \
    && yarn --version \
# Print version and verify bashrc is properly sourced so that everything works in the interactive shell
    && bash -c "node --version" \
    && bash -c "npm --version" \
    && bash -c "yarn --version"


WORKDIR /home/frappe
EXPOSE 8000 9000 6787
```

</details>

**Comment by revant** [comment]
> Weird. What OS are you on?

Linux.

In container `frappe` user has 999 as uid
User on docker host is 1000. 

Right now image does this with

```
RUN groupadd -g 1000 frappe \
  && useradd -ms /bin/bash -u 1000 -g 1000 -G sudo frappe \
  && chown -R 1000:1000 /home/frappe \
```

**Comment by ddavidebor** [comment]
It is not ideal in this scenario, because it will increase build time and complexity if the user wants to fiddle with the image, while saving no more than a few seconds when downloading.

Development images should put the developer first, not the machine.

**Comment by revant** [comment]
> It is not ideal in this scenario, because it will increase build time and complexity if the user wants to fiddle with the image, while saving no more than a few seconds when downloading.
> 
> Development images should put the developer first, not the machine.

I don't mind big dev images.

but uid and permission main is blocker

**Comment by revant** [comment]
> Weird. What OS are you on?

Adding user like this solved my problem

```
RUN groupadd -g 1000 frappe \
    && useradd --no-log-init -r -m -u 1000 -g 1000 -G sudo frappe
```

Also,

yarn is only installed in node v12. current image installs yarn in both v10 and v12

**Comment by revant** [comment]
@Thunderbottom can confirm permissions error?

**Comment by ddavidebor** [comment]
@revant now yarn is installed in both versions, and UID and GID are set to 1000, can merge
Verified everything still works normally on mac, no permission errors now

**Pull request merged**